man/dbfilter.1
.TH DBFILTER 1 " 3 April 1996"
.FB
.SH NAME
dbfilter \- filter an existing \fIFirstBase\fP index into a
new \fIFirstBase\fP index
.SH SYNOPSIS
.B dbfilter
[\fB-b\fP]\ [\fB-y\fP]\
[\fB-s\fP]\ [\fB-a\fP]\
[\fB-d\fP\ \fIdatabase\fP]
[\fB-i\fP\ \fIindex\fP]
[\fB-c\fP\ \fIcolumns\fP]
[\fB-f\fP\ \fIfilter_index\fP]
[\fIfilter_file\fP | \fB-\fP ]
.PD
.SH DESCRIPTION
.I Dbfilter
generates a distinct new index using an existing \fIindex\fP
and a \fIfilter_file\fP.
Each line of the \fIfilter_file\fP, a standard text file,
is looked up in the \fIindex\fP.
If the referenced index entry exists, it is written to the new
\fIfilter_index\fP. In other words, the \fIfilter_file\fP is used to filter
\fIindex\fP into \fIfilter_index\fP.
.PP
The result, \fIfilter_index\fP, is a complete \fIFirstBase\fP index.
(Note: since \fIdbigen\fP(1) is not used to generate \fIfilter_index\fP,
no \fI*.idicti\fP file is used). Error messages are generated for
each \fIfilter_file\fP request that cannot be referenced.
Using \fB-a\fP finds all occurrences of each filter line, and
\fB-s\fP sorts \fIfilter_index\fP after completion.
.PP
The \fB-c\fP switch allows setting the number of character columns or
positions that will be used to generate the index search key.
Normally, the width of the index indicates the number of character columns
required to match during the index lookup. The \fIfilter_file\fP can be
wider than this \(em only the first \fIcolumns\fP are used to reference the
indexed record. The \fIcolumns\fP can also be set to less than the width of
the index.
.PP
The \fB-b\fP flag will run the \fIFirstBase\fP tool in batch mode,
suitable for shell scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed.
.PP
If the \fIfilter_file\fP specified is \fB-\fP, standard input (the keyboard)
is read, and batch mode (\fB-b\fP) is set. This works also when batch mode
is set, but no \fIfilter_file\fP is specified. For example the following
is valid:
.br
.nf
.nj
.ft CW
grep pattern text_file | dbfilter -d dbase -i index -f new -c 10 -b
.fi
.ad
.PP
Another interesting use of \fIdbfilter\fP can be achieved by generating
an index (\fIdbigen\fP(1)) and using \fIdbawk\fP(1), \fIdbemit\fP(1) or
\fIdbmerge\fP(1) to generate a \fIfilter_file\fP. This \fIfilter_file\fP
can then be edited or filtered using other Unix tools (\fIsed\fP(1),
\fIawk\fP(1), etc). The \fIfilter_file\fP is then used to generate
a new \fIFirstBase\fP index, usable with existing tools and scripts.
Remember to
make the first part of the \fIfilter_file\fP look like the first part of
the indexed field.
.PP
.I Dbfilter(1)
provides a convenient method of filtering an existing \fIFirstBase\fP index
into a new index that can then be used anywhere in the \fIFirstBase\fP system.
.SH FILES
.PD 0
.TP 10
dbase
default name of database.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary file.
.TP 10
FILTER
default \fIfilter_file\fP.
.TP 10
filter
default \fIfilter_index\fP.
.PD
.SH SEE ALSO
dbdind(1), dbigen(1), dbawk(1), dbemit(1)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbfilter\fP(1)
will ask permission before overwriting an existing \fIFirstBase\fP index
when not run in batch mode.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbfilter.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbfilter.c,v 9.2 2001/02/16 19:47:51 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbfilter_sid[] = "@(#) $Id: dbfilter.c,v 9.2 2001/02/16 19:47:51 john Exp $";
#endif

/* 
 *  dbfilter.c - target is a standard Cdb index based on a text
 *	file. Look up the lines of the loadfile, generate an
 *	index entry for these. Report on those not found.
 */

#include <fb.h>
#include <fb_vars.h>

#if !FB_PROTOTYPES
extern main();
static int filter();
static int getindex();
static int nextindex();
static FILE *finit();
void usage();
#else /* FB_PROTOTYPES */
extern main(int, char **);
static int filter(FILE *);
static int getindex(char *, fb_database *);
static int nextindex(char *, fb_database *);
static FILE *finit(int, char **);
void usage(void);
#endif /* FB_PROTOTYPES */

static int vflag = 0;			/* verbose flag */
static int tflag = 0;			/* trace flag */
static int sflag = 0;			/* sflag - sort flag */
static int aflag = 0;			/* aflag - all matches flag */
static int column = 0;			/* column cut off for matching */
static int ksize = 0;			/* strlen of key */
static char filter_fb_basename[FB_MAXNAME];
static char filter_idx[FB_MAXNAME];
static char filter_idict[FB_MAXNAME];
static int idx_fd;
static int idict_fd;

#define FILTER_BASENAME	"filter"
#define FILTER_LOADNAME	"FILTER"


/*
 *  dbfilter - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         char str[FB_MAXLINE];
         FILE *fs;


         (void) Dbfilter_sid;

	 if (fb_testargs(argc, argv, "-") > 0)
            cdb_batchmode = 1;	/* force cdb_batchmode if using stdio */
         fb_getargs(argc, argv, FB_ALLOCDB);
	 fb_opendb(cdb_db, READ, FB_WITHINDEX, FB_MUST_INDEX);
	 fb_allow_int();		/* allows instant piping */	
         fb_scrhdr(cdb_db, "Open Files"); fb_infoline();
	 fs = finit(argc, argv);
	 if (!cdb_batchmode && !cdb_yesmode){
	    sprintf(str, "Ok to Generate Filter %s (y/n) ?",
               filter_fb_basename);
	    if (fb_mustbe('y', str, cdb_t_lines, 1) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
         if (filter(fs) == FB_ERROR)
	    fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 fclose(fs);
	 fb_closedb(cdb_db);
	 fb_ender();
      }

/* 
 *  filter - filter all lines from a load file (fs) into a Cdb index
 */
 
   static int filter(fs)
      FILE *fs;
      
      {
         long n = 0L;
         char line[FB_MAXLINE], *p, key[FB_MAXLINE], msg[FB_MAXLINE];
         char com[FB_MAXLINE];

         if (!cdb_batchmode){
	    FB_XWAIT();
	    fb_gcounter(n);
	    }
         /* init target filter index */

	 for (;;){
            if (fgets(line, FB_MAXLINE, fs) == NULL)
               break;
            if ((p = strchr(line, FB_NEWLINE)) != 0)
               *p = NULL;
            line[column] = NULL;
            strcpy(key, line);
            if (FB_OFNUMERIC(cdb_db->ip[0]->type))
	       fb_makess(key, cdb_db->ip[0]->type, cdb_db->ip[0]->size);
            ksize = strlen(key);
            if (tflag && cdb_batchmode)
               fb_serror(FB_MESSAGE, key, NIL);
            if (getindex(key, cdb_db) != FB_AOK){
               fb_serror(FB_MESSAGE, "Could not find: ", key);
               continue;
               }

            for (;;){
               n++;
               if (!cdb_batchmode)
                  fb_gcounter(n);
               fb_nextwrite(0, cdb_db->irec);	/* enter into filter index */
               if (!aflag)
                  break;
               /* aflag is set here, so find the next match */
               if (nextindex(key, cdb_db) != FB_AOK)
                  break;
               }
	    }
	 fb_wflush(1);
         close(idx_fd);
	 fb_putxhead(idict_fd, n, n);
         close(idict_fd);

         if (sflag){
            if (!cdb_batchmode){
               fb_scrstat("Sorting");
               fb_move(4,1), fb_clrtobot();
               sprintf(msg, "Sorting %s", filter_fb_basename);
               fb_fmessage(msg);
               }
            sprintf(com, "sort -o %s %s", filter_idx, filter_idx);
            fb_system(com, FB_WITHROOT);
            }
	 if ((!cdb_batchmode && !cdb_yesmode)){
	    if (!cdb_batchmode)
	       fb_fmessage(NIL);
	    sprintf(com, "There were %ld Records Selected", n);
	    fb_serror(FB_MESSAGE, com, NIL);
	    }
         return(FB_AOK);
      }

/*
 *  getindex - try and get an index record loaded
 *	this is fb_getxrec(3) without the getrec tied in.
 */

   static getindex(s, dp)
      char *s;
      fb_database *dp;
      
      {
         long rec;
	 int st = FB_ERROR;
         char *p;
	 
         /* b_tree handled if needed */
         if (dp->b_tree){
            rec = fb_btree_search(s, dp->b_idx, dp->b_seq);
            if (rec > 0 && rec <= dp->reccnt)
               st = FB_AOK;
            p = fb_key_ptr(dp->b_seq);
            strcpy(dp->irec, p);
            dp->irec[dp->irecsiz-1] = FB_NEWLINE;
            }
         /* standard FB index */
         else{
            rec = fb_megasearch(dp->ifd, s, 0, 1, dp->bsend, dp->bsmax,
               dp->irecsiz, 1, dp->irec);
            if (rec > 0 && rec <= dp->bsmax){
               rec = atol((char *) 
                  (dp->irec + dp->irecsiz - 11));	/* FB_RECORDPTR + 1 */
               if (rec > 0 && rec <= dp->reccnt)
                  st = FB_AOK;
               }
            }
	 return(st);
      }

/*
 *  nextindex - force the next indexed record to be loaded.
 */

   static int nextindex(s, dp)
      char *s;
      fb_database *dp;
      
      {
         long rec;
	 int st = FB_ERROR;
         fb_bseq *bs;
         char *p;
	 
         if (dp->b_tree){				/* btree path */
            bs = dp->b_seq;
            /*
             * the current node has to be re-read in case a key was deleted
             * by some other process.
             */
            /*
            if (bs->bs_curkey >= 1 && bs->bs_curkey <= 3)
               fb_seq_getrec(bs->bs_recno, bs);
            */
            for (;;){
               if (bs->bs_curkey >= 3){
                  if (bs->bs_next == 0)
                     return(st);
                  fb_seq_getrec(bs->bs_next, bs);
                  bs->bs_curkey = 0;
                  }
               bs->bs_curkey++;
               rec = fb_key_eval(bs);
               if (rec == 0)
                  bs->bs_curkey = 4;
               else if (rec > 0L && rec <= dp->reccnt)	/* ignore rec of 0 */
                  break;
               }
            p = fb_key_ptr(bs);
            if (strncmp(p, s, ksize) == 0){
               strcpy(dp->irec, p);
               dp->irec[dp->irecsiz-1] = FB_NEWLINE;
               st = FB_AOK;
               }
            }
         else {
            while (read(dp->ifd, dp->irec, (unsigned) dp->irecsiz) ==
                  dp->irecsiz){
               dp->bsrec += 1L;
               rec = atol((char *) (dp->irec + dp->irecsiz - 11));
               if (rec > 0L && rec <= dp->reccnt){	/* ignore rec of 0 */
                  if (strncmp(s, dp->irec, ksize) == 0)
                     st = FB_AOK;
                  break;
                  }
               }
            }
	  return(st);
      }

/*
 *  finit - intialize --- merely check for existance.
 */
 
   static FILE *finit(argc, argv)
      int argc;
      char *argv[];
      
      {
         char str[FB_MAXLINE], fname[FB_MAXNAME], seq[FB_SEQSIZE + 1];
         char buf[FB_MAXLINE];
	 FILE *fs;
         long bsmax, bsorg;
	 int i;
	 
	 fname[0] = NULL;
         strcpy(filter_fb_basename, FILTER_BASENAME);
         
	 for (i = 1; i < argc; i++)
	    if (argv[i][0] == '-'){		/* skip dbase stuff */
	       if (argv[i][1] == 'd' || argv[i][1] == 'i')
	          i++;
	       else if (argv[i][1] == 't')
                  tflag = 1;
	       else if (argv[i][1] == 'a')
                  aflag = 1;
	       else if (argv[i][1] == 'v')
                  vflag = 1;
	       else if (argv[i][1] == 's')
                  sflag = 1;
	       else if (argv[i][1] == 'c'){	/* get column indicator */
                  if (++i >= argc)
                     usage();
	          column = atoi(argv[i]);
                  if (column >= FB_MAXLINE)
                     fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE],
                        "Column Wrong");
                  }
	       else if (argv[i][1] == 'f'){	/* get filter name */
                  if (++i >= argc)
                     usage();
	          strcpy(filter_fb_basename, argv[i]);
                  }
	       }
	    else{
	       strcpy(fname, argv[i]);
	       break;
	       }
         if (column == 0)
            column = cdb_db->irecsiz;
         /* build target filter (index) names */
         sprintf(filter_idx,
               SYSMSG[S_FMT_2S], filter_fb_basename, SYSMSG[S_EXT_IDX]);
         sprintf(filter_idict,
               SYSMSG[S_FMT_2S], filter_fb_basename, SYSMSG[S_EXT_IDICT]);
	 if (equal(filter_idx, cdb_db->dindex))
	    fb_xerror(FB_ABORT_ERROR,
               "Filter_index MUST be distinct from index", NIL);
	 if (access(filter_idx, 0) != -1 && !cdb_batchmode && !cdb_yesmode){
	    sprintf(str, 
	       "Permission to OVERWRITE filter `%s' (y = yes, <cr> = no)? ",
               filter_fb_basename);
	    if (fb_mustbe('y', str, 15, 10) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
         
	 if (fname[0] == NULL){		/* requires full path name */
            if (cdb_batchmode == 0){
               strcpy(str, "FILTER");
               if (fb_getfilename(fname, "Filter Lookup File: ", str) ==
                     FB_ERROR)
                  fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
               }
	    }
         if (fname[0] != NULL){
            sprintf(str, "from %s", fname);
            fb_scrstat(str);
            fs = fb_mustfopen(fname, "r");
            }
         else
            fs = stdin;

         /* lastly, create the new filter (index) files */
         close(creat(filter_idx, 0666));
         close(creat(filter_idict, 0666));
         if ((idict_fd = open(filter_idict, READWRITE)) < 0)
            fb_xerror(FB_CANT_OPEN, filter_idict, NIL);
         sprintf(seq, SYSMSG[S_FMT_04D], fb_getseq(cdb_db->fd));
         lseek(idict_fd, FB_SEQSTART, 0);		/* write SEQF */
         write(idict_fd, seq, FB_SEQSIZE);
         fb_putxhead(idict_fd, 0L, 0L);		/* positions fds also */
         fb_r_init(cdb_db->ihfd);
         fb_getxhead(cdb_db->ihfd, &bsmax, &bsorg);
         for(; fb_nextline(buf, FB_MAXLINE) != 0;){
            if (buf[0] == '%')
               break;
            strcat(buf, "\n");
            write(idict_fd, buf, strlen(buf));
            }
         close(cdb_db->ihfd);
         close(idict_fd);

	 idict_fd = open(filter_idict, READWRITE);
	 if ((idx_fd = open(filter_idx, READWRITE)) < 0)
	    fb_xerror(FB_CANT_OPEN, filter_idx, NIL);

	 fb_w_init(1, idx_fd, -1);

         return(fs);
      }

/* 
 *  usage message 
 */
 
   void usage()
      {
         fb_xerror(FB_MESSAGE,
	    "usage: dbfilter [-d dbase] [-i index] [filter options] file",NIL);
      }
